### PR TITLE
my your dads

### PR DESCRIPTION
**Pull request by denizbeser**

### CODE REVIEW

File: tests/language_specific/english/test_english_language_generator.py [hunk 1]
```diff
@@ -245,6 +246,132 @@ def test_dad_put_a_cookie_in_a_box_using_i():
     ).as_token_sequence() == ("I", "put", "a", "cookie", "in", "a", "box")
 
 
+def test_dad_put_a_cookie_in_a_box_using_you():
```
Comment (rgabbard) [inline tests/language_specific/english/test_english_language_generator.py]: We also want an test for the simple case of "Dad has a cookie" with no other action going on.

File: adam/language_specific/english/english_language_generator.py [hunk 2]
```diff
@@ -193,7 +194,38 @@ def _translate_object_to_noun(
             if (dependency_node.part_of_speech != PROPER_NOUN) and (
                 MASS_NOUN not in lexicon_entry.properties
             ):
-                if count == 1:
+
+                # Objects that might possess the noun
+                try:
```
Comment (rgabbard) [inline adam/language_specific/english/english_language_generator.py]: I know it is supposedly "Pythonic", but I am not a fan of using exceptions for flow control.

I would suggest rewriting this section in the following way:
* first check if anything has a `HAS` relation with the object in question.
* if it does, then examine that particular object to see how to render it as a word.

File: adam/language_specific/english/english_language_generator.py [hunk 3]
```diff
@@ -193,7 +194,34 @@ def _translate_object_to_noun(
...
+                    # If speaker possesses the noun
+                    if IS_SPEAKER in possessor.properties:
+                        quantifier_node = DependencyTreeToken("my", DETERMINER)
+                        quantifier_role = DETERMINER_ROLE
+                    # If addressee possess the noun
+                    elif IS_ADDRESSEE in possessor.properties:
+                        quantifier_node = DependencyTreeToken("your", DETERMINER)
+                        quantifier_role = DETERMINER_ROLE
+                    # otherwise, if the agent possess the noun
+                    elif possessor == only(
```
Comment (rgabbard) [inline adam/language_specific/english/english_language_generator.py]: Actually, I think the desired behavior is to use _'s_ if a non-agent is the possessor; if the agent is a possessor, just fall back on the regular determiner behavior.  

Also this is getting complex enough we should pull it out into its own `add_determiner` method.

File: adam/language_specific/english/english_language_generator.py [hunk 4]
```diff
@@ -220,6 +209,63 @@ def _translate_object_to_noun(
 
             return dependency_node
 
+        def add_determiner(
+            self,
+            _object: SituationObject,
+            count: int,
+            dependency_node: DependencyTreeToken,
```
Comment (rgabbard) [inline adam/language_specific/english/english_language_generator.py]: Maybe rename to `noun_dependency_node` so it is clear which node you are talking about

File: adam/language_specific/english/english_language_generator.py [hunk 5]
```diff
@@ -220,6 +209,63 @@ def _translate_object_to_noun(
...
+                elif (
+                    len(possession_relations) == 1
+                    and possession_relations[0].first_slot
+                    not in only(self.situation.actions).argument_roles_to_fillers[AGENT]
+                ):
+                    token = (
+                        self.objects_to_dependency_nodes[
+                            possession_relations[0].first_slot
+                        ].token
+                        + "'s"
```
Comment (rgabbard) [inline adam/language_specific/english/english_language_generator.py]: I think most dependency trees analyze `'s` as its own token, attached to the possessor by a `CASE` role.

File: adam/language_specific/english/english_language_generator.py [hunk 6]
```diff
@@ -220,6 +209,63 @@ def _translate_object_to_noun(
...
+            if len(possession_relations) > 1:
+                raise RuntimeError("Can not handle multiple possession relations")
+            else:
+                # If speaker possesses the noun
+                if (
+                    len(possession_relations) == 1
+                    and IS_SPEAKER in possession_relations[0].first_slot.properties
+                ):
+                    quantifier_node = DependencyTreeToken("my", DETERMINER)
+                    quantifier_role = DETERMINER_ROLE
```
Comment (rgabbard) [inline adam/language_specific/english/english_language_generator.py]: for possessives the role should be `nmod:poss`

File: tests/language_specific/english/test_english_language_generator.py [hunk 7]
```diff
@@ -270,6 +273,97 @@ def test_dad_put_a_cookie_in_a_box_using_you():
...
+
+
+def test_dad_has_a_cookie():
+    dad = SituationObject(DAD)
+    cookie = SituationObject(COOKIE)
+    situation = HighLevelSemanticsSituation(
+        ontology=GAILA_PHASE_1_ONTOLOGY,
+        objects=[dad, cookie],
+        relations=[],
+        actions=[SituationAction(HAVE, ((AGENT, dad), (THEME, cookie)))],
```
Comment (rgabbard) [inline tests/language_specific/english/test_english_language_generator.py]: `HAVE` should not be an available action type (remove it from the `Ontology` if it is there). Rather, if there is a `has` relation and no action, then translate to English using _has_

File: adam/language_specific/english/english_language_generator.py [hunk 8]
```diff
@@ -220,6 +219,66 @@ def _translate_object_to_noun(
...
+                    determiner_node = DependencyTreeToken("your", DETERMINER)
+                    determiner_role = DETERMINER_ROLE
+                # add 's if a non-agent possess the noun
+                elif (
+                    len(possession_relations) == 1
+                    and self.situation.is_dynamic
+                    and possession_relations[0].first_slot
+                    not in only(self.situation.actions).argument_roles_to_fillers[AGENT]
+                ):
+                    token = self.objects_to_dependency_nodes[
```
Comment (rgabbard) [inline adam/language_specific/english/english_language_generator.py]: Needs test covered (which, as you note, requires translating the possessing object)

File: adam/language_specific/english/english_language_generator.py [hunk 9]
```diff
@@ -331,6 +390,54 @@ def _preposition_for_region_as_goal(self, region: Region[SituationObject]) -> st
                     f"Don't know how to translate {region} to a preposition yet"
                 )
 
+        def _translate_relation_to_verb(
+            self, relation: Relation[SituationObject]
+        ) -> DependencyTreeToken:
+            lexicon_entry = self._unique_lexicon_entry(relation.relation_type)
+            if any(
```
Comment (rgabbard) [inline adam/language_specific/english/english_language_generator.py]: Is there similar logic for other verbs that we can refactor into a method?

File: adam/language_specific/english/english_language_generator.py [hunk 10]
```diff
@@ -331,6 +390,54 @@ def _preposition_for_region_as_goal(self, region: Region[SituationObject]) -> st
...
+            )
+            self.dependency_graph.add_node(verb_dependency_node)
+
+            first_slot_dependency_node = self.objects_to_dependency_nodes[
+                relation.first_slot
+            ]
+            self.dependency_graph.add_edge(
+                first_slot_dependency_node,
+                verb_dependency_node,
+                role=self._translate_argument_role(AGENT),
```
Comment (rgabbard) [inline adam/language_specific/english/english_language_generator.py]: I would just specify the roles to be subject and direct object directly; no need to do the translation step.

File: adam/language_specific/english/english_language_generator.py (hunk 9 above)
Comment (denizbeser) [reply adam/language_specific/english/english_language_generator.py]: That's slightly different because we access those through argument roles.

Comment (rgabbard) [reply adam/language_specific/english/english_language_generator.py]: Got it.

**Pull request merged**